Sources/ProcedureObserver.swift

```diff
@@ -0,0 +1,148 @@
+//
+//  ProcedureKit
+//
+//  Copyright © 2016 ProcedureKit. All rights reserved.
+//
+
+import Foundation
+
+/**
+ Types which conform to this protocol, can be attached to `Procedure` subclasses to receive
+ events at state transitions.
+ */
+public protocol ProcedureObserver {
+
+    /**
+     Observer gets notified when it is attached to a procedure.
+
+     - parameter procedure: the observed `Procedure`.
+     */
+    func didAttachTo(procedure: Procedure)
+}
+
+public extension ProcedureObserver {
+
+    /**
+     Default implementation does nothing.
+
+     - parameter procedure: the observed `Procedure`.
+     */
+    func didAttachTo(procedure: Procedure) { }
```

Comment:
I think a more Swift 3 name would be `didAttach(to procedure: Procedure)`, so the use site would look like `didAttach(to: procedure)`. Also, for the other protocol names, it might be more fitting for them to drop the `procedure` external parameter label. So `willExecute(procedure: Procedure)` would be come `willExecute(_ procedure: Procedure)`, as that would allow usage like `willExecute(procedure)`.